File: src/common/response.rs
use serde::Serialize;

/// Standard success response envelope: `{ "data": T }`
#[derive(Debug, Serialize)]
pub struct ApiResponse<T: Serialize> {
    pub data: T,
}

impl<T: Serialize> ApiResponse<T> {
    pub fn success(data: T) -> Self {
        Self { data }
    }
}

/// Paginated collection response:
/// `{ "data": [...], "meta": { ... }, "links": { ... } }`
#[derive(Debug, Serialize)]
pub struct PaginatedResponse<T: Serialize> {
    pub data: Vec<T>,
    pub meta: PaginationMeta,
    pub links: PaginationLinks,
}

#[derive(Debug, Serialize)]
pub struct PaginationMeta {
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
    pub total_pages: i64,
}

#[derive(Debug, Serialize)]
pub struct PaginationLinks {
    #[serde(rename = "self")]
    pub self_link: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prev: Option<String>,
    pub first: String,
    pub last: String,
}

impl<T: Serialize> PaginatedResponse<T> {
    /// Build a paginated response from a list of items and pagination info.
    ///
    /// `base_path` should be the path portion of the URL, e.g. `/api/v1/users`.
    pub fn new(data: Vec<T>, total: i64, page: i64, per_page: i64, base_path: &str) -> Self {
        let total_pages = if total == 0 {
            1
        } else {
            (total + per_page - 1) / per_page
        };

        let self_link = format!("{}?page={}&per_page={}", base_path, page, per_page);
        let first = format!("{}?page=1&per_page={}", base_path, per_page);
        let last = format!("{}?page={}&per_page={}", base_path, total_pages, per_page);

        let next = if page < total_pages {
            Some(format!(
                "{}?page={}&per_page={}",
                base_path,
                page + 1,
                per_page
            ))
        } else {
            None
        };

        let prev = if page > 1 {
            Some(format!(
                "{}?page={}&per_page={}",
                base_path,
                page - 1,
                per_page
            ))
        } else {
            None
        };

        Self {
            data,
            meta: PaginationMeta {
                total,
                page,
                per_page,
                total_pages,
            },
            links: PaginationLinks {
                self_link,
                next,
                prev,
                first,
                last,
            },
        }
    }
}

File: webapp/src/http/response.ts
/**
 * HTTP 响应相关的类型定义和工具函数
 */

/**
 * 通用的 API 成功响应结构: { data: T }
 */
export interface ApiResponse<T = unknown> {
  data: T;
}

/**
 * 分页响应元信息
 */
export interface PaginationMeta {
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

/**
 * 分页响应链接
 */
export interface PaginationLinks {
  self: string;
  next?: string;
  prev?: string;
  first: string;
  last: string;
}

/**
 * 分页响应数据: { data: T[], meta: PaginationMeta, links: PaginationLinks }
 */
export interface PaginatedResponse<T = unknown> {
  data: T[];
  meta: PaginationMeta;
  links: PaginationLinks;
}

/**
 * API 错误响应体
 */
export interface ApiErrorResponse {
  error: {
    code: string;
    message: string;
    details?: FieldError[];
  };
}

/**
 * 字段级验证错误
 */
export interface FieldError {
  field: string;
  message: string;
  code: string;
}

/**
 * HTTP 错误响应
 */
export interface HttpError {
  code: string;
  message: string;
  data?: unknown;
  status?: number;
}

/**
 * API 业务错误，携带后端返回的错误码和附加数据
 */
export class ApiError extends Error {
  readonly code: string;
  readonly data: unknown;

  constructor(code: string, message: string, data?: unknown) {
    super(message);
    this.name = "ApiError";
    this.code = code;
    this.data = data;
  }
}

/**
 * 从 ApiResponse<T> 信封中提取 data 字段。
 * 如果传入的值不包含 data 字段，直接返回原值（兼容直接返回的简单值）。
 */
export const unwrapData = <T>(response: ApiResponse<T> | T): T => {
  if (response != null && typeof response === "object" && "data" in response) {
    return (response as ApiResponse<T>).data;
  }
  return response as T;
};
